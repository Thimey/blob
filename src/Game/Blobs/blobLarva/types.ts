import { ActorRefFrom, StateMachine } from 'xstate';
import { Coordinates, PersistedActor, BlobSpawn } from 'src/types';

export interface Context {
  id: string;
  position: Coordinates;
  destination: Coordinates;
  larvaHeadRadius: number;
  larvaBodyRadiusX: number;
  larvaBodyRadiusY: number;
  pupa?: {
    spawnTo: 'bloblet';
    spawnTime: number;
    hatchAt: number;
  };
}

export type DrawEvent = {
  type: 'DRAW';
  ctx: CanvasRenderingContext2D;
};

export type DrawLarvaSelectedEvent = {
  type: 'DRAW_LARVA_SELECTED';
  ctx: CanvasRenderingContext2D;
};

export type UpdateEvent = {
  type: 'UPDATE';
};

export type LarvaClickEvent = {
  type: 'LARVA_CLICKED';
  id: string;
};

export type LarvaSpawnSelected = {
  type: 'LARVA_SPAWN_SELECTED';
  blobToSpawn: BlobSpawn;
  spawnTime: number;
  hatchAt: number;
};

export type PupaHatch = {
  type: 'PUPA_HATCH';
};

type StateValues = 'larva' | 'pupa' | 'hatched';

export type State = {
  value: StateValues;
  context: Context;
};

export type Events =
  | DrawEvent
  | UpdateEvent
  | LarvaClickEvent
  | LarvaSpawnSelected
  | DrawLarvaSelectedEvent
  | PupaHatch;

export type BlobLarvaActor = ActorRefFrom<StateMachine<Context, any, Events>>;
export type PersistedLarvaActor = PersistedActor<Context, string[]>;
