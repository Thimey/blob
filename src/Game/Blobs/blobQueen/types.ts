import { ActorRefFrom, StateMachine } from 'xstate';
import { Point, DrawEvent, UpdateEvent } from 'game/types';

export type { DrawEvent, UpdateEvent };

export interface Context {
  position: Point;
  bodyRadiusX: number;
  bodyRadiusY: number;
  eyeRadiusX: number;
  eyeRadiusY: number;
  eyeOffsetX: number;
  eyeOffsetY: number;
}

type StateValues = 'idle' | 'blinkingClose' | 'blinkingOpen';

export type State = {
  value: StateValues;
  context: Context;
};

export type Event = DrawEvent | UpdateEvent;

export type BlobQueenActor = ActorRefFrom<StateMachine<Context, any, Event>>;
