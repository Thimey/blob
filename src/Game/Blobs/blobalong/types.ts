import { ActorRefFrom, StateMachine } from 'xstate';
import {
  Movement,
  Point,
  DrawEvent,
  UpdateEvent,
  MapClickEvent,
} from 'game/types';
import { Connection } from 'game/blobNetwork';

export interface Context {
  id: string;
  position: Point;
  rotation: number;
  finRotation: number;
  finRotationDir: 1 | -1;
  movement?: Movement;
  makingConnection?: {
    connection: Connection;
    growPoints: Point[];
    currentPointIndex: number;
    head1Rotation: number;
    head2Rotation: number;
    newEndNodeCentre?: Point;
    newEndNodePointAngle?: number;
    newEndNodeGrowStartAngle?: number;
    newEndNodeGrowEndAngle?: number;
  };
}

export type BlobalongClickEvent = {
  type: 'BLOBALONG_CLICK';
  id: string;
};

export type DrawSelectedEvent = {
  type: 'DRAW_SELECTED';
  ctx: CanvasRenderingContext2D;
};

export type MakeConnectionEvent = {
  type: 'MAKE_CONNECTION';
  connection: Connection;
  growPoints: Point[];
  newEndNodeCentre?: Point;
};

export type Event =
  | DrawEvent
  | UpdateEvent
  | MapClickEvent
  | BlobalongClickEvent
  | DrawSelectedEvent
  | MakeConnectionEvent;

export type BlobalongActor = ActorRefFrom<StateMachine<Context, any, Event>>;
