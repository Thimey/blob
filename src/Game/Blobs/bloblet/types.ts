import { ActorRefFrom, StateMachine } from 'xstate';
import {
  Point,
  PersistedActor,
  DrawEvent,
  UpdateEvent,
  MapClickEvent,
  Movement,
  SelectEvent,
  DeselectEvent,
  MultiSelectEvent,
} from 'game/types';

export type { UpdateEvent, DrawEvent, MapClickEvent, Movement };

export type Context = {
  id: string;
  position: Point;
  radius: number;
  movement?: Movement;
  harvestingShrub?: {
    startAt: number;
    shrubId: string;
    harvestRate: number;
    position: Point;
    leafPositions: Point[];
    amount: number;
  };
};

export type BlobletClickEvent = {
  type: 'BLOBLET_CLICKED';
  id: string;
};

export type ShrubClickEvent = {
  type: 'SHRUB_CLICKED';
  clickCoordinates: Point;
  shrubPosition: Point;
  shrubId: string;
  harvestRate: number;
  leafPositions: Point[];
  amount: number;
};

export type DrawSrubEvent = {
  type: 'DRAW_SHRUB';
  ctx: CanvasRenderingContext2D;
};

export type DrawSelectedEvent = {
  type: 'DRAW_SELECTED';
  ctx: CanvasRenderingContext2D;
};

export type ShrubDepletedEvent = {
  type: 'SHRUB_DEPLETED';
  shrubId: string;
};

export type StateValues =
  | { selection: 'deselected' }
  | { selection: 'selected' }
  | { movement: 'stationary' }
  | { movement: 'moving' }
  | { movement: { harvestingShrub: 'movingToShrub' } }
  | { movement: { harvestingShrub: 'atShrub' } }
  | { movement: { harvestingShrub: 'movingToQueen' } }
  | { movement: { harvestingShrub: 'atQueen' } };

export type State = {
  value: StateValues;
  context: Context;
};

export type Event =
  | BlobletClickEvent
  | MapClickEvent
  | SelectEvent
  | DeselectEvent
  | MultiSelectEvent
  | DrawEvent
  | UpdateEvent
  | ShrubClickEvent
  | ShrubDepletedEvent
  | DrawSelectedEvent
  | DrawSrubEvent;

export type BlobletActor = ActorRefFrom<StateMachine<Context, any, Event>>;
export type PersistedBlobletActor = PersistedActor<Context, string[]>;
