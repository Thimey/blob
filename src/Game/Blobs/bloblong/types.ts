import { ActorRefFrom, StateMachine } from 'xstate';
import {
  Movement,
  Point,
  DrawEvent,
  UpdateEvent,
  MapClickEvent,
} from 'game/types';

export interface Context {
  id: string;
  position: Point;
  rotation: number;
  movement?: Movement;
}

type StateValues =
  | { selection: 'deselected' }
  | { selection: 'selected' }
  | { movement: 'stationary' }
  | { movement: 'moving' };

export type State = {
  value: StateValues;
  context: Context;
};

export type BloblongClickEvent = {
  type: 'BLOBLONG_CLICK';
  id: string;
};

export type DrawSelectedEvent = {
  type: 'DRAW_SELECTED';
  ctx: CanvasRenderingContext2D;
};

export type Event =
  | DrawEvent
  | UpdateEvent
  | MapClickEvent
  | BloblongClickEvent
  | DrawSelectedEvent;

export type BloblongActor = ActorRefFrom<StateMachine<Context, any, Event>>;
