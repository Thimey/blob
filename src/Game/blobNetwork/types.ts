import { Point } from 'game/types';

export interface Node {
  id: string;
  centre: Point;
  radiusX: number;
  radiusY: number;
  connections: {
    [nodeId: NodeId]: {
      connectionId: Connection['id'];
      direction: 'startToEnd' | 'endToStart';
    };
  };
  overlappingNodes: NodeId[];
}

export interface Connection {
  id: string;
  start: Point;
  end: Point;
  bezierP1: Point;
  bezierP2: Point;
  points: Point[];
}

export type NodeId = Node['id'];
export type NodeMap = Record<NodeId, Node>;

export type ConnectionId = Connection['id'];
export type ConnectionMap = Record<ConnectionId, Connection>;

export interface Network {
  nodes: NodeMap;
  connections: ConnectionMap;
}

export type PathStrategy = 'networkOnly' | 'linear';
