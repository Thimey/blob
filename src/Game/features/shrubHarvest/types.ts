import { Interpreter } from 'xstate';
import { UpdateEvent, DrawEvent, ClickedEvent } from 'game/types';
import { ShrubActor, PersistedShrubActor } from 'game/resources/shrub';
import { BlobletActor, PersistedBlobletActor } from './bloblet';

export interface Context {
  bloblets: BlobletActor[];
  shrubs: ShrubActor[];
}

export type HarvestShrubEvent = {
  type: 'HARVEST_SHRUB';
  shrubId: string;
  harvestCount: number;
};

export type FeedOnShrubEvent = {
  type: 'FEED_SHRUB';
  amount: number;
};

export type ShrubDepletedEvent = {
  type: 'SHRUB_DEPLETED';
  shrubId: string;
};

export type GrowShrubEvent = {
  type: 'GROW_SHRUB';
};

export type PersistedGameState = {
  bloblets: PersistedBlobletActor[];
  shrubs: PersistedShrubActor[];
} & Omit<Context, 'bloblets' | ' shrubs'>;

type StateValues = 'initialising' | 'ready';

export type State = {
  value: StateValues;
  context: Context;
};

export type Event =
  | DrawEvent
  | UpdateEvent
  | ClickedEvent
  | FeedOnShrubEvent
  | HarvestShrubEvent
  | ShrubDepletedEvent
  | GrowShrubEvent;

export type BlobQueenService = Interpreter<Context, any, Event, State>;
