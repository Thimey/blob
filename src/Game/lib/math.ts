import { Point } from '../types';

export function generateId() {
  return Date.now().toString();
}

export function roundTo(number: number, decimalPlaces: number) {
  return decimalPlaces < 1
    ? Math.round(number)
    : Math.round(number * 10 * decimalPlaces) / (10 * decimalPlaces);
}

export function closestToZero(x1: number, x2: number) {
  return Math.abs(x1) < Math.abs(x2) ? x1 : x2;
}

export function getDistance({ x: x1, y: y1 }: Point, { x: x2, y: y2 }: Point) {
  return Math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2);
}

export function makeRandNumber(min = 0, max = 1) {
  return Math.random() * (max - min + 1) + min;
}

export type RGB = { r: number; g: number; b: number };
const blackRGB: RGB = { r: 0, g: 0, b: 0 };
export function hexToRGB(hex: string): RGB {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);

  if (result) {
    return {
      r: parseInt(result[1], 16),
      g: parseInt(result[2], 16),
      b: parseInt(result[3], 16),
    };
  }

  return blackRGB;
}

interface Ellipse {
  x: number;
  y: number;
  radiusX: number;
  radiusY: number;
}

export function isPointWithinEllipse(ellipse: Ellipse, { x, y }: Point) {
  return (
    (x - ellipse.x) ** 2 / ellipse.radiusX ** 2 +
      (y - ellipse.y) ** 2 / ellipse.radiusY ** 2 <=
    1
  );
}

export function isPointWithinCircle(
  position: Point,
  radius: number,
  mousePosition: Point
) {
  const distanceFromClick = getDistance(position, mousePosition);

  return distanceFromClick <= radius;
}

interface Rectangle {
  x: number;
  y: number;
  width: number;
  height: number;
}

export function isPointWithinRectangle(rect: Rectangle, { x, y }: Point) {
  return (
    x >= rect.x &&
    x <= rect.x + rect.width &&
    y >= rect.y &&
    y <= rect.y + rect.height
  );
}

export function makeCubicBezierPoints(
  p0: Point,
  p1: Point,
  p2: Point,
  p3: Point,
  step: number
) {
  const cx = 3 * (p1.x - p0.x);
  const bx = 3 * (p2.x - p1.x) - cx;
  const ax = p3.x - p0.x - cx - bx;

  const cy = 3 * (p1.y - p0.y);
  const by = 3 * (p2.y - p1.y) - cy;
  const ay = p3.y - p0.y - cy - by;

  let t = 0;
  const points: Point[] = [];

  while (t <= 1) {
    const x = ax * t ** 3 + bx * t ** 2 + cx * t + p0.x;
    const y = ay * t ** 3 + by * t ** 2 + cy * t + p0.y;

    points.push({ x, y });
    t += 1 / step;
  }

  return points;
}

export function makeLinearPoints(start: Point, end: Point, step = 1) {
  const totalDistance = getDistance(start, end);
  const dx = end.x - start.x;
  const dy = end.y - start.y;
  const angle = Math.atan(dy / dx);

  let distance = totalDistance - step;
  const points = [start];
  while (distance > 0) {
    const x = distance * Math.cos(angle);
    const y = distance * Math.sin(angle);

    points.push({ x, y });
    distance -= Math.min(step, distance);
  }
  points.push(end);

  return points;
}
