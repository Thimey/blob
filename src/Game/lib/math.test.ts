import { Point } from 'game/types';
import { makeLinearPoints } from './math';

// function printLine(points: Point[]) {
//   const line = points.map(({ x, y }) => ());
// }

describe('math', () => {
  describe('makeLinearPath', () => {
    it('should get correct path', () => {
      /*
        end(0, 0)                
        \
         \
          \
           \
            start(4, 3)
      */

      expect(
        makeLinearPoints({ x: 3, y: 3 }, { x: 0, y: 0 }, 1)
      ).toMatchSnapshot();
    });
  });
});
