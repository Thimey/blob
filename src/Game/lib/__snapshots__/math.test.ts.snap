// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`math makeLinearPath should get correct path 1`] = `
Array [
  Object {
    "x": 3,
    "y": 3,
  },
  Object {
    "x": 2.2928932188134525,
    "y": 2.292893218813452,
  },
  Object {
    "x": 1.5857864376269049,
    "y": 1.5857864376269046,
  },
  Object {
    "x": 0.8786796564403572,
    "y": 0.8786796564403571,
  },
  Object {
    "x": 0.17157287525380965,
    "y": 0.17157287525380963,
  },
  Object {
    "x": 0,
    "y": 0,
  },
]
`;
